### PR TITLE
[Spot] Auto file_mounts translation for spot

### LINKED ISSUE #1073: [Spot] Automatically translate the local file mounts and workdir to sky Storage
The friction of changing an original available sky yaml with `workdir` and local file mounts set to the managed spot yaml has been observed multiple times. We may want to have SkyPilot translate those for the users automatically, i.e. translate those into SkyPilot Storage with `persistent: false` and `mode: COPY`

### PR DESCRIPTION
**Pull request by Michaelvll**
Fixes #1073.

This PR makes sure that a normal task YAML can be run with `sky spot launch` without any modification.

~~It is currently blocked by #1069.~~

Tested:
- [x] `sky spot launch -n test-storage exampels/managed_spot_with_storage.yaml`

TODO:
- [x] Add test

### CODE REVIEW

File: sky/execution.py [hunk 1]
```diff
@@ -377,25 +378,44 @@ def spot_launch(
               f'{colorama.Style.RESET_ALL}')
         return
 
-    # TODO(zhwu): Refactor the Task (as Resources), so that we can enforce the
-    # following validations.
-    # Check the file mounts in the task.
-    # Disallow all local file mounts (copy mounts).
+    # Translate the workdir and local file mounts to cloud file mounts.
+    new_storage_mounts = {}
```
Comment (romilbhardwaj) [inline sky/execution.py:382]: One thing to think about would be how are ctrl+c/failure events handled for these newly created storage objects. For example, consider the case when a user ctrl+c `sky spot launch` after the storage has been uploaded but before the spot controller is launched. Correct me if I am wrong, but is there a chance the storage object would not get automatically deleted since teardown might not be run?

The good news is it will still be logged in `sky storage ls`, but we may have to point users to manually delete these objects. #1028 might also reduce chances of this happening.

**Review by romilbhardwaj — COMMENTED** [review]
Thanks the draft of this PR!

File: sky/execution.py (hunk 1 above)
Comment (Michaelvll) [reply sky/execution.py:382]: That is a great point! Any errors during the run could cause the storage leakage mentioned. Another problem here would be even the spot job finished and removed the storage, the `sky storage ls` on the laptop will still contain that storage. 
We may want to have a `--refresh` option for the `sky storage ls` as well.

File: sky/utils/common_utils.py [hunk 2]
```diff
@@ -9,12 +9,23 @@
...
 
 from sky import sky_logging
 
 logger = sky_logging.init_logger(__name__)
 
+_run_id = None
+
+
+def get_run_id():
+    """Returns a unique run id for this logging."""
```
Comment (concretevitamin) [inline sky/utils/common_utils.py:23]: Update doc?
- rm 'for this logging'
- perhaps expand on the definition: e.g., does the run id preserve across a spot job's recovery attempts?

File: sky/spot/constants.py [hunk 3]
```diff
@@ -7,3 +7,6 @@
 SPOT_CONTROLLER_YAML_PREFIX = '~/.sky/spot_controller'
 
 SPOT_TASK_YAML_PREFIX = '~/.sky/spot_tasks'
+
+SPOT_WORKDIR_BUCKET_NAME = 'sky-spot-workdir-{username}-{hash}'
+SPOT_FILE_MOUNT_BUCKET_NAME = 'sky-spot-fm-{username}-{hash}'
```
Comment (concretevitamin) [inline sky/spot/constants.py:12]: nits:
- `fm` -> `file-mounts` / `filemounts`?  As buckets are user-facing, this may be easier to spot.
- (for discussion) wdyt about starting to use `skypilot` for such new constants?

File: sky/execution.py [hunk 4]
```diff
@@ -377,31 +378,52 @@ def spot_launch(
...
+        )
+        new_storage_mounts[backend_utils.SKY_REMOTE_WORKDIR] = (
+            storage.Storage.from_yaml_config({
+                'name': bucket_name,
+                'source': workdir,
+                'persistent': False,
+                'mode': 'COPY',
+            }))
+        logger.info(
+            f'Translated workdir {workdir} to cloud storage {bucket_name}.')
```
Comment (concretevitamin) [inline sky/execution.py:400]: How about 
> Workdir {workdir} will be synced to cloud storage {bucket_name}. See `sky storage ls`.

Not too sure about how to hint the last part as at this point it won't show this Storage object?

File: sky/execution.py [hunk 5]
```diff
@@ -377,31 +378,52 @@ def spot_launch(
...
+        new_storage_mounts[dst] = storage.Storage.from_yaml_config({
+            'name': bucket_name,
+            'source': src,
+            'persistent': False,
+            'mode': 'COPY',
+        })
+        logger.info(
+            f'Translated local file mount {dst} to cloud storage {bucket_name}.'
+        )
+    task.update_storage_mounts(new_storage_mounts)
```
Comment (concretevitamin) [inline sky/execution.py:421]: Very unlikely: but it will be good to check the existing storage mounts don't take up these keys.

File: sky/execution.py [hunk 6]
```diff
@@ -377,31 +378,52 @@ def spot_launch(
...
-                f'jobs.\nFound local source paths:\n{local_sources}\nHint: use '
-                'Storage to auto-upload local files to a cloud bucket.')
+    # TODO(zhwu): Optimize this by:
+    # 1. Use the same bucket for all the mounts.
+    # 2. When the src is the same, use the same bucket.
+    for i, (dst, src) in enumerate(copy_mounts.items()):
+        task.file_mounts.pop(dst)
+        bucket_name = spot.constants.SPOT_FILE_MOUNT_BUCKET_NAME.format(
+            username=getpass.getuser(),
+            hash=f'{run_id}-{i}',
```
Comment (concretevitamin) [inline sky/execution.py:410]: perhaps it's better to `hash` -> `id`?

**Review by concretevitamin — COMMENTED** [review]
This will be a big improvement to spot.  A quick pass.

File: sky/data/storage.py [hunk 7]
```diff
@@ -193,7 +193,7 @@ def download_remote_dir(self, local_path: str) -> None:
                 continue
             path = os.path.join(local_path, remote_path)
             if not os.path.exists(os.path.dirname(path)):
-                os.makedirs(os.path.dirname(path))
+                os.makedirs(os.path.dirname(path), exist_ok=True)
```
[the remark this answers is not in the document]
Comment (Michaelvll) [reply sky/data/storage.py]: Good catch! Fixed. Thanks!

File: sky/utils/common_utils.py (hunk 2 above)
Comment (Michaelvll) [reply sky/utils/common_utils.py]: Updated the docs. For the unique id for across a spot job's recovery attempts, I will have another PR adding that ;)

File: sky/spot/constants.py (hunk 3 above)
Comment (Michaelvll) [reply sky/spot/constants.py]: Good point! Fixed. Thanks!

File: sky/execution.py (hunk 4 above)
Comment (Michaelvll) [reply sky/execution.py]: Good point! I moved the `sky storage ls` hint before the storage upload, so that the user knows that the storage will be shown in the table after the source is uploaded.

File: sky/execution.py (hunk 6 above)
Comment (Michaelvll) [reply sky/execution.py]: Done. Thanks!

File: sky/execution.py (hunk 5 above)
Comment (Michaelvll) [reply sky/execution.py:472]: Good point! Added the check for the files only mounts. For other parts, the destination should be already checked when the task is constructed.

File: sky/task.py [hunk 8]
```diff
@@ -485,13 +494,42 @@ def update_storage_mounts(self, storage_mounts: Dict[str,
...
+            resources = list(self.resources)[0]
+            storage_cloud = resources.cloud
+            if storage_cloud is None:
+                # Get the first enabled cloud.
+                enabled_clouds = global_user_state.get_enabled_clouds()
+                if len(enabled_clouds) == 0:
+                    check.check(quiet=True)
+                    enabled_clouds = global_user_state.get_enabled_clouds()
+                if len(enabled_clouds) == 0:
+                    raise ValueError('No enabled clouds.')
```
Comment (concretevitamin) [inline sky/task.py:518]: Refactor this into a helper? So that it's only possible to trigger a state change if L530 holds.

actually/in addition, can refactor this whole thing from L497 to `Task.get_preferred_cloud_for_storage()`.

File: sky/task.py [hunk 9]
```diff
@@ -579,6 +617,12 @@ def set_file_mounts(self, file_mounts: Optional[Dict[str, str]]) -> None:
                             f'File mount source {source!r} does not exist '
                             'locally. To fix: check if it exists, and correct '
                             'the path.')
+            if (target == constants.SKY_REMOTE_WORKDIR and
+                    self.workdir is not None):
+                with ux_utils.print_exception_no_traceback():
+                    raise ValueError(
+                        f'Cannot mount to {constants.SKY_REMOTE_WORKDIR!r} '
```
Comment (concretevitamin) [inline sky/task.py:624]: nit: Cannot use {constants.SKY_REMOTE_WORKDIR!r} as a destination path of a file mount, as it will be used by the workdir.

Actually, why don't we always allow it (so if workdir is specified, users ensure conflict-freeness; this may be useful for uploading some local files outside of local workdir to the remote workdir) -- the status quo --, or always ban it?  I find the conditional ban a bit strange, because a second `sky launch` on an alive cluster may conditionally trigger this.

File: sky/utils/common_utils.py [hunk 10]
```diff
@@ -20,7 +20,7 @@
 
 
 def get_run_id():
-    """Returns a unique run id for this logging."""
+    """Returns a unique run id for each run of the program."""
```
Comment (concretevitamin) [inline sky/utils/common_utils.py:23]:
```
    """Returns a unique run id for each 'run'.

    A run is defined as the lifetime of a process that has imported `sky` and has called its CLI or programmatic APIs. For example, two successive `sky launch` are two runs."""
```
Should define "run" as it's non-obvious.  Is the "or programmatic APIs" part accurate?

File: sky/data/storage.py [hunk 11]
```diff
@@ -45,6 +48,29 @@ class StorageMode(enum.Enum):
...
+            raise ValueError(f'Unknown cloud type: {cloud}')
+
+
+def get_store_prefix(storetype: StoreType) -> str:
+    if storetype == StoreType.S3:
+        return 's3://'
+    elif storetype == StoreType.GCS:
+        return 'gs://'
+    elif storetype == StoreType.AZURE:
+        return 'azure://'
```
Comment (concretevitamin) [inline sky/data/storage.py:69]: omit?

File: sky/data/storage.py [hunk 12]
```diff
@@ -45,6 +48,29 @@ class StorageMode(enum.Enum):
     COPY = 'COPY'
 
 
+def get_storetype_from_cloud(cloud: clouds.Cloud) -> StoreType:
+    if isinstance(cloud, clouds.AWS):
+        return StoreType.S3
+    elif isinstance(cloud, clouds.GCP):
+        return StoreType.GCS
+    elif isinstance(cloud, clouds.Azure):
+        return StoreType.AZURE
```
Comment (concretevitamin) [inline sky/data/storage.py:57]: omit? (and change L60 message)? otherwise, `add_store(StoreType.AZURE)` could've been called.

File: sky/execution.py [hunk 13]
```diff
@@ -378,55 +379,120 @@ def spot_launch(
...
         return
 
+    # ================================================================
     # Translate the workdir and local file mounts to cloud file mounts.
-    new_storage_mounts = {}
+    # ================================================================
     run_id = common_utils.get_run_id()[:8]
 
+    # Step 1: Translate the workdir SkyPilot storage.
+    logger.info(f'{colorama.Fore.YELLOW}Translating local file_mounts '
```
Comment (concretevitamin) [inline sky/execution.py:388]: nit: "Translating local file_mounts" -> "Translating file_mounts with local source paths"

File: sky/execution.py [hunk 14]
```diff
@@ -378,55 +379,120 @@ def spot_launch(
               f'{colorama.Style.RESET_ALL}')
         return
 
+    # ================================================================
     # Translate the workdir and local file mounts to cloud file mounts.
-    new_storage_mounts = {}
+    # ================================================================
```
Comment (concretevitamin) [inline sky/execution.py:384]: UX nits: I feel L366 and L369-370 can be omitted. The former is strongly implied. The latter currently has 1 strategy only. Wdyt?

File: sky/execution.py [hunk 15]
```diff
@@ -378,55 +379,120 @@ def spot_launch(
...
                 'source': workdir,
                 'persistent': False,
                 'mode': 'COPY',
-            }))
-        logger.info(
-            f'Translated workdir {workdir} to cloud storage {bucket_name}.')
-
-    copy_mounts = task.get_local_to_remote_file_mounts()
+            })
+        # Check of the existance of the workdir in file_mounts is done in
```
Comment (concretevitamin) [inline sky/execution.py:404]: nit: existence

File: sky/execution.py [hunk 16]
```diff
@@ -378,55 +379,120 @@ def spot_launch(
               f'{colorama.Style.RESET_ALL}')
         return
 
+    # ================================================================
```
Comment (concretevitamin) [inline sky/execution.py:382]: Is it possible / easy to refactor L382 -- end of translation to a helper? If not, no worries.

File: sky/execution.py [hunk 17]
```diff
@@ -378,55 +379,120 @@ def spot_launch(
...
+                'source': local_fm_path,
+                'persistent': False,
+                'mode': 'MOUNT',
+            })
+        if spot.constants.SPOT_FM_REMOTE_TMP_DIR in original_storage_mounts:
+            with ux_utils.print_exception_no_traceback():
+                raise ValueError(
+                    'Failed to translate file mounts, due to the default '
+                    f'destination {spot.constants.SPOT_FM_REMOTE_TMP_DIR} '
+                    'is taken.')
```
Comment (concretevitamin) [inline sky/execution.py:466]: nit: is -> being

File: sky/execution.py [hunk 18]
```diff
@@ -378,55 +379,120 @@ def spot_launch(
...
     # so we need to copy the files to the bucket manually here before sending to
     # the remote spot controller.
+    logger.info(f'{colorama.Fore.YELLOW}Uploading storage from sources...'
+                f'{colorama.Style.RESET_ALL} See sky storage ls')
     task.add_storage_mounts()
 
-    # Replace the source field that is local path in all storage_mounts with
-    # bucket URI and remove the name field.
+    # Step 5: Add the file download into the file mounts, such as
+    #  /origin-dst: s3://spot-fm-file-only-bucket-name/file-0
```
Comment (concretevitamin) [inline sky/execution.py:483]: nit: origin -> original

File: sky/execution.py [hunk 19]
```diff
@@ -378,55 +379,120 @@ def spot_launch(
...
+        if spot.constants.SPOT_FM_REMOTE_TMP_DIR in original_storage_mounts:
+            with ux_utils.print_exception_no_traceback():
+                raise ValueError(
+                    'Failed to translate file mounts, due to the default '
+                    f'destination {spot.constants.SPOT_FM_REMOTE_TMP_DIR} '
+                    'is taken.')
+        sources = list(src_to_file_id.keys())
+        sources_str = '\n\t'.join(sources)
+        logger.info(
+            f'Sources with file will be synced to cloud storage {bucket_name}:'
```
Comment (concretevitamin) [inline sky/execution.py:470]: nit: Source files in file_mounts will be ...

File: sky/execution.py [hunk 20]
```diff
@@ -378,55 +379,120 @@ def spot_launch(
...
+            f'\n\t{sources_str}')
     task.update_storage_mounts(new_storage_mounts)
 
+    # Step 4: Upload storage from sources
     # Copy the local source to a bucket. The task will not be executed locally,
     # so we need to copy the files to the bucket manually here before sending to
     # the remote spot controller.
+    logger.info(f'{colorama.Fore.YELLOW}Uploading storage from sources...'
```
Comment (concretevitamin) [inline sky/execution.py:478]: nit: Uploading sources to cloud storage

File: sky/execution.py [hunk 21]
```diff
@@ -378,55 +379,120 @@ def spot_launch(
...
+        sources_str = '\n\t'.join(sources)
+        logger.info(
+            f'Sources with file will be synced to cloud storage {bucket_name}:'
+            f'\n\t{sources_str}')
     task.update_storage_mounts(new_storage_mounts)
 
+    # Step 4: Upload storage from sources
     # Copy the local source to a bucket. The task will not be executed locally,
     # so we need to copy the files to the bucket manually here before sending to
     # the remote spot controller.
```
Comment (concretevitamin) [inline sky/execution.py:477]: Move L475-477?

File: sky/execution.py [hunk 22]
```diff
@@ -378,55 +379,120 @@ def spot_launch(
...
+    logger.info(f'{colorama.Fore.YELLOW}Uploading storage from sources...'
+                f'{colorama.Style.RESET_ALL} See sky storage ls')
     task.add_storage_mounts()
```
Comment (concretevitamin) [inline sky/execution.py:480]: @romilbhardwaj I've always been confused by this func. Can we rename it to point out it's not lazy? `sync_storage_mounts()`? Similarly `add_store()` is syncing up.

File: sky/spot/constants.py [hunk 23]
```diff
@@ -8,5 +8,8 @@
 
 SPOT_TASK_YAML_PREFIX = '~/.sky/spot_tasks'
 
-SPOT_WORKDIR_BUCKET_NAME = 'sky-spot-workdir-{username}-{hash}'
-SPOT_FILE_MOUNT_BUCKET_NAME = 'sky-spot-fm-{username}-{hash}'
+SPOT_WORKDIR_BUCKET_NAME = 'skypilot-spot-workdir-{username}-{id}'
+SPOT_FM_BUCKET_NAME = 'skypilot-spot-filemounts-folder-{username}-{id}'
```
Comment (concretevitamin) [inline sky/spot/constants.py:12]: folder -> folders

File: sky/spot/constants.py [hunk 24]
```diff
@@ -8,5 +8,8 @@
 
 SPOT_TASK_YAML_PREFIX = '~/.sky/spot_tasks'
 
-SPOT_WORKDIR_BUCKET_NAME = 'sky-spot-workdir-{username}-{hash}'
-SPOT_FILE_MOUNT_BUCKET_NAME = 'sky-spot-fm-{username}-{hash}'
+SPOT_WORKDIR_BUCKET_NAME = 'skypilot-spot-workdir-{username}-{id}'
```
Comment (concretevitamin) [inline sky/spot/constants.py:11]: 1. L11-14 have sky and skypilot. Should we be consistent?
2. Consider removing the "spot" word. When we add serverless I imagine these buckets can be reused.  Wdyt?

File: sky/task.py [hunk 25]
```diff
@@ -464,13 +467,19 @@ def set_storage_mounts(
             self.storage_mounts = None
             return self
         for target, _ in storage_mounts.items():
+            if (target == constants.SKY_REMOTE_WORKDIR and
+                    self.workdir is not None):
+                with ux_utils.print_exception_no_traceback():
+                    raise ValueError(
+                        f'Cannot mount to {constants.SKY_REMOTE_WORKDIR!r} '
```
Comment (concretevitamin) [inline sky/task.py:474]: ditto

File: sky/task.py [hunk 26]
```diff
@@ -485,13 +494,42 @@ def update_storage_mounts(self, storage_mounts: Dict[str,
 
     def add_storage_mounts(self) -> None:
```
Comment (concretevitamin) [inline sky/task.py:495]: See the other comment about renaming this. @romilbhardwaj 

(This is actually uploading things. Its name should reflect that to warn callers.)

File: sky/task.py (hunk 8 above)
Comment (concretevitamin) [reply sky/task.py:518]: Should also add a TODO to say such placement decisions should be refactored and put into optimizer instead.

**Review by concretevitamin — COMMENTED** [review]
Did a complete pass. Some comments.

File: sky/execution.py (hunk 22 above)
Comment (romilbhardwaj) [reply sky/execution.py:480]: I agree - this needs to be changed in the (now long pending) Storage [refactor](https://github.com/skypilot-org/skypilot/discussions/1028)..

File: sky/task.py (hunk 26 above)
Comment (romilbhardwaj) [reply sky/task.py:495]: Yes, let's change the name for now and maybe add a todo stating that this should be lazy

**Comment by Michaelvll** [comment]
Tested the `sky spot launch --cloud gcp examples/using_file_mounts.yaml` with the following changes:
```
15c15
< workdir: .
---
> workdir: ./examples
17c17
< num_nodes: 2
---
> num_nodes: 1
29c29
<   /sky: .
---
>   /sky: ./examples
55c55
<   /tmp/workdir: ~/tmp-workdir
---
>   # /tmp/workdir: ~/tmp-workdir
78,79c78,79
<   sudo apt update
<   sudo apt install tree
---
>   # sudo apt update
>   # sudo apt install tree
92c92
<   tree /tmp/workdir
---
>   # tree /tmp/workdir
99c99,100
<   pip3 install -e .
---
>   # pip3 install -e .
>   ls minimal.yaml
118c119,120
<   ! ls sky.egg-info/
---
>   # ! ls sky.egg-info/
>   ls minimal.yaml
122,123c124,125
<   ls /tmp/workdir/circle-link
<   ! readlink -e /tmp/workdir/circle-link
---
>   # ls /tmp/workdir/circle-link
>   # ! readlink -e /tmp/workdir/circle-link
```

File: sky/backends/backend_utils.py [hunk 27]
```diff
@@ -1909,7 +1909,8 @@ def interrupt_handler(signum, frame):
     logger.warning(f'{colorama.Fore.LIGHTBLACK_EX}The job will keep '
                    f'running after Ctrl-C.{colorama.Style.RESET_ALL}')
     kill_children_processes()
-    raise KeyboardInterrupt(exceptions.KEYBOARD_INTERRUPT_CODE)
+    with ux_utils.print_exception_no_traceback():
```
Comment (concretevitamin) [inline sky/backends/backend_utils.py:1912]: Nice!

File: sky/execution.py [hunk 28]
```diff
@@ -380,20 +380,70 @@ def spot_launch(
...
+def _translate_local_file_mounts(task: task_lib.Task) -> task_lib.Task:
     # ================================================================
     # Translate the workdir and local file mounts to cloud file mounts.
     # ================================================================
+    task = copy.deepcopy(task)
     run_id = common_utils.get_run_id()[:8]
+    original_file_mounts = task.file_mounts if task.file_mounts else {}
+    original_storage_mounts = task.storage_mounts if task.storage_mounts else {}
 
     # Step 1: Translate the workdir SkyPilot storage.
```
Comment (concretevitamin) [inline sky/execution.py:432]: nit: "to" after workdir?

File: sky/execution.py (hunk 21 above)
Comment (concretevitamin) [reply sky/execution.py]: I was confused by original L467-478 here
```
    # Copy the local source to a bucket. The task will not be executed locally,
    # so we need to copy the files to the bucket manually here before sending to
    # the remote spot controller.
```
on second look, they do describe `task.sync_storage_mounts()` line below?  If so, I think we should change "copy" to "upload"/"sync" to avoid confusion.

File: sky/utils/ux_utils.py [hunk 29]
```diff
@@ -17,6 +17,6 @@ def print_exception_no_traceback():
                 raise ValueError('...')
     """
     original_tracelimit = getattr(sys, 'tracebacklimit', 1000)
-    sys.tracebacklimit = 0
+    # sys.tracebacklimit = 1
```
Comment (concretevitamin) [inline sky/utils/ux_utils.py:20]: Remnant?

File: sky/data/storage.py [hunk 30]
```diff
@@ -54,7 +56,8 @@ def get_storetype_from_cloud(cloud: clouds.Cloud) -> StoreType:
     elif isinstance(cloud, clouds.GCP):
         return StoreType.GCS
     elif isinstance(cloud, clouds.Azure):
-        return StoreType.AZURE
+        with ux_utils.print_exception_no_traceback():
+            raise ValueError('Azure is not supported yet.')
```
Comment (concretevitamin) [inline sky/data/storage.py:60]: Azure Blob Storage is not...

File: sky/data/storage.py [hunk 31]
```diff
@@ -66,9 +69,11 @@ def get_store_prefix(storetype: StoreType) -> str:
     elif storetype == StoreType.GCS:
         return 'gs://'
     elif storetype == StoreType.AZURE:
-        return 'azure://'
+        with ux_utils.print_exception_no_traceback():
+            raise ValueError('Azure is not supported yet.')
```
Comment (concretevitamin) [inline sky/data/storage.py:73]: ditto

File: sky/data/storage.py [hunk 32]
```diff
@@ -836,7 +841,6 @@ def sync_local_dir(self) -> None:
                     if not line:
                         break
                     str_line = line.decode('utf-8')
-                    logger.info(str_line)
```
Comment (concretevitamin) [inline sky/data/storage.py:813]: Why this change?

File: sky/task.py [hunk 33]
```diff
@@ -471,8 +470,11 @@ def set_storage_mounts(
                     self.workdir is not None):
```
Comment (concretevitamin) [inline sky/task.py:470]: Is `target == constants.SKY_REMOTE_WORKDIR` sufficient?  The RHS is a relative path starting with `~/`. Should we catch the case where `target` equals `/home/ubuntu/sky_workdir` assuming `ubuntu` is the username?

File: sky/task.py [hunk 34]
```diff
@@ -525,9 +527,14 @@ def add_storage_mounts(self) -> None:
         if storage_cloud is None:
             raise ValueError('No available cloud to mount storage.')
         store_type = storage_lib.get_storetype_from_cloud(storage_cloud)
+        return store_type
+
+    def sync_storage_mounts(self) -> None:
+        """Adds storage mounts to the Task."""
```
Comment (concretevitamin) [inline sky/task.py:533]: Syncs storage mounts: sync files/dirs to cloud storage.

File: sky/task.py [hunk 35]
```diff
@@ -550,12 +557,6 @@ def add_storage_mounts(self) -> None:
                         mnt_path: blob_path,
                     })
                 elif store_type is storage_lib.StoreType.GCS:
-                    # Remember to run `gcloud auth application-default login`
```
Comment (concretevitamin) [inline sky/task.py:506]: Any change made this no longer needed?

File: sky/clouds/gcp.py [hunk 36]
```diff
@@ -18,6 +18,11 @@
     '~/.config/gcloud/'
     'application_default_credentials.json')
 
+GCP_CONFIGURE_PATH = '~/.config/gcloud/configurations/config_default'
+# Do not place the backup under the gcloud config directory, as ray
+# autoscaler can overwrite that directory on the remote nodes.
+GCP_CONFIGURE_SKY_BACKUP_PATH = '~/.sky/.sky_gcp_config_default'
```
Comment (concretevitamin) [inline sky/clouds/gcp.py:24]: nit: rename CONFIGURE -> CONFIG as the former can be parsed as a verb.

File: sky/cloud_stores.py [hunk 37]
```diff
@@ -93,26 +94,17 @@ class GcsCloudStorage(CloudStorage):
...
-        'pushd /tmp &>/dev/null',
-        '(test -f ~/google-cloud-sdk/bin/gsutil || (wget --quiet '
-        'https://dl.google.com/dl/cloudsdk/channels/rapid/downloads/'
-        'google-cloud-sdk-367.0.0-linux-x86_64.tar.gz && '
-        'tar xzf google-cloud-sdk-367.0.0-linux-x86_64.tar.gz && '
-        'mv google-cloud-sdk ~/ && '
-        '~/google-cloud-sdk/install.sh -q ))',
-        'popd &>/dev/null',
-    ]
+    _GET_GSUTIL = gcp.GCLOUD_INSTALLATION_COMMAND
```
Comment (concretevitamin) [inline sky/cloud_stores.py:97]: Strikes me as odd - is it true that the RHS installs `gsutil`? Worth a comment.

File: sky/clouds/gcp.py [hunk 38]
```diff
@@ -29,6 +34,21 @@
 
 _IMAGE_ID_PREFIX = ('projects/deeplearning-platform-release/global/images/')
 
+_GCLOUD_INSTALLATION_LOG = '~/.sky/logs/gcloud_installation.log'
+# Need to be run with /bin/bash
+GCLOUD_INSTALLATION_COMMAND = f'pushd /tmp &>/dev/null && \
```
Comment (concretevitamin) [inline sky/clouds/gcp.py:39]: Can we add a comment on why we factored this out?

File: sky/clouds/gcp.py [hunk 39]
```diff
@@ -328,13 +348,32 @@ def check_credentials(self) -> Tuple[bool, Optional[str]]:
         return True, None
 
     def get_credential_file_mounts(self) -> Dict[str, str]:
+        # Create a backup of the config_default file, as the original file can
```
Comment (concretevitamin) [inline sky/clouds/gcp.py:351]: Nit: say this backup file will be uploaded to the remote clusters and applied appropriately.

File: sky/authentication.py [hunk 40]
```diff
@@ -212,27 +208,12 @@ def setup_gcp_authentication(config):
...
-            subprocess.run(f'cp {config_path} {sky_backup_config_path}',
-                           shell=True,
-                           check=True)
-        new_file_mounts = config.get('file_mounts', {})
-        new_file_mounts[
-            GCP_CONFIGURE_SKY_BACKUP_PATH] = GCP_CONFIGURE_SKY_BACKUP_PATH
-        config['file_mounts'] = new_file_mounts
+            clouds.gcp.GCP_CONFIGURE_SKY_BACKUP_PATH)
+        assert os.path.exists(sky_backup_config_path
+                             ), 'GCP credential backup file does not exist.'
```
Comment (concretevitamin) [inline sky/authentication.py:215]: To be safe, add a hint  ".. check local path {sky_backup_config_path} exists."?

**Review by concretevitamin — COMMENTED** [review]
Looks good @Michaelvll - Mostly nits.

> It is currently blocked by https://github.com/skypilot-org/skypilot/issues/1069.

Is this still accurate?

File: sky/cloud_stores.py (hunk 37 above)
Comment (Michaelvll) [reply sky/cloud_stores.py:99]: Added the comment. Thanks!

File: sky/clouds/gcp.py (hunk 38 above)
Comment (Michaelvll) [reply sky/clouds/gcp.py:41]: Added! Thanks!

File: sky/clouds/gcp.py (hunk 39 above)
Comment (Michaelvll) [reply sky/clouds/gcp.py:353]: Added. Thanks!

File: sky/data/storage.py (hunk 30 above)
Comment (Michaelvll) [reply sky/data/storage.py]: Fixed. Thanks!

File: sky/data/storage.py (hunk 31 above)
Comment (Michaelvll) [reply sky/data/storage.py]: Fixed. Thanks!

File: sky/data/storage.py (hunk 32 above)
Comment (Michaelvll) [reply sky/data/storage.py:813]: The gsutil cp will generate a lot of progress output to the stderr causing the console filled with a large amount of outputs. I feel it would be better to not show those outputs. The stderr will be shown in the L857 below if the error occurs.

File: sky/execution.py (hunk 28 above)
Comment (Michaelvll) [reply sky/execution.py]: Done. Thanks!

File: sky/task.py (hunk 34 above)
Comment (Michaelvll) [reply sky/task.py]: Good catch! Fixed. Thanks!

File: sky/task.py (hunk 35 above)
Comment (Michaelvll) [reply sky/task.py:506]: These lines are not effective as the file_mounts will be executed before the setup commands are executed. Instead, I changed the `_GSUTIL` in `cloud_store.py` to `_GSUTIL = (f'GOOGLE_APPLICATION_CREDENTIALS='{gcp.DEFAULT_GCP_APPLICATION_CREDENTIAL_PATH} gsutil')`, so that the credential path is added to the environment variable.

File: sky/utils/ux_utils.py (hunk 29 above)
Comment (Michaelvll) [reply sky/utils/ux_utils.py]: Good catch! Restored the original code. Thanks!

**Review by Michaelvll — COMMENTED** [review]
Thanks for the review @concretevitamin ! The note is not valid now, as I applied a workaround  for the files to be uploaded (hard link the files to a folder and upload the folder to the bucket).

File: sky/clouds/gcp.py (hunk 36 above)
Comment (Michaelvll) [reply sky/clouds/gcp.py]: Done. Thanks!

File: sky/execution.py (hunk 21 above)
Comment (Michaelvll) [reply sky/execution.py]: Ahh, good point! Just replace the `copy` with `upload`. Thanks!

File: tests/test_smoke.py [hunk 41]
```diff
@@ -282,7 +282,7 @@ def test_job_queue():
     test = Test(
         'job_queue',
         [
-            f'sky launch -y -c {name} examples/job_queue/cluster.yaml',
+            f'sky launch --cloud gcp -y -c {name} examples/job_queue/cluster.yaml',
```
Comment (concretevitamin) [inline tests/test_smoke.py:285]: Shall we leave these out?

**Review by concretevitamin — APPROVED** [review]
LGTM - this will simplify the usage of managed spot a lot @Michaelvll !

File: sky/task.py (hunk 33 above)
Comment (Michaelvll) [reply sky/task.py:472]: Good catch! It is a bit tricky to catch the case. I added a TODO comment and leave it to the future. : )

**Comment by Michaelvll** [comment]
Tested:
- [ ] `tests/run_smoke_tests.sh`
- [ ] `sky spot launch -n test-storage examples/managed_spot_with_storage.yaml`

File: tests/test_smoke.py (hunk 41 above)
Comment (Michaelvll) [reply tests/test_smoke.py]: Done. Thanks!
